### ISSUE TITLE
Auto enable like RANCiD ?

### ISSUE THREAD

**Issue by FTBZ** [ytti/oxidized#213]
I didn't find a way to not provide the enable password in oxidized like it's in RANCiD? There's a way to do that?

**Comment by ytti** [comment]
Not sure I understand. Just not configure enable and it won't be used.

**Comment by FTBZ** [comment]
Sorry I'm wrong. I thought that my user was in "auto enable" on the Cisco device but it's not the case. I'm using a specific set of "privilege exec" for this user. This role based method is working with RANCiD and seems to be not working with Oxidized.

I noticed that the RANCiD "rancid.types.conf" is taking about this and uses the full command as workaround (show running-config view full).

An idea if I need to create a custom model or if it's something that will be fixed?

**Comment by ytti** [comment]
I'm not sure I fully understand your problem. If you're using RBAC you can allow 'show run' command, and no parameters should be needed.

If there is particular reason why this should work, we probably could hack something in the model to support both use cases. Maybe something in line with


cfg @@sh_run do |cfg|
   #parse_show_running_config
end

````
cfg :telnet, :ssh do
  @@sh_run  = "show running-configuration"
  @@sh_run += " view full" if vars("some_variable_blaa_blaa")
  ..
....
````

**Comment by FTBZ** [comment]
Based on my experience with RANCiD, when using RBAC it's not enough to allow only "show run". It will only loads these informations without "enabled mode":

>Current configuration : 183 bytes
>!
>! Last configuration change at 06:45:26 MET Thu Oct 8 2015 by p22893
>! NVRAM config last updated at 17:10:21 MET Wed Oct 7 2015
>!
>boot-start-marker
>boot-end-marker
>!
>!
>!
>!
>!
>!
>end

I'm not a code expert, but I will test your workaround.

**Comment by ytti** [comment]
When you login to the box normally with the credentials you have, and you run 'show run', do you get full configuration? If not, then we need to change the code or you need to change the config. 

I know that you can configure IOS RBAC so that 'show running-config' will display full config.

**Comment by FTBZ** [comment]
Nope, with my settings the "sh run" only shows the configuration pasted in the preceding post. I need to specify the full "show running-config view all" to get the configuration.

I appreciate the time you take, I will promote this project on my side.

**Comment by ytti** [comment]
Can you show me your RBAC configuration?

**Comment by FTBZ** [comment]
I made many tests using "parser view" and the more simple "privilege exec" commands. Nothing seems to be working. After searching the Cisco/RANCID forums during hours, I've found always the same response:
> ...you have to issue show running-config view full instead of show running-config to see current running config.

This is probably why it's included natively in RANCID. There is a way to simply add this to Oxidized in the next release?

**Comment by ytti** [comment]
You could try the suggestion I gave if it works. Might be easier for testing purposes to copy ios.rb to ~/.config/oxidized/model/ios.rb and modify that version. 

Idea is that we check variable and depending on variable define what command to give for running-config. Variable could be set in router.db like this.

router.db:
router1:ios:rbac
router2:junos
router3:ios
router4:ios:rbac

Then in .config/oxidized/config you'd have
```
source:
  cvs:
    vars_map:
      ios_rbac: 2
```

Now in ios.rb you could use vars('ios_rbac') to verify its existence.

**Comment by jkrauska** [comment]
Agreed -- 'view full' would be a nicer default.

This problem comes up when creating a user with limited access rights on ios.
(so that the oxidized user cannot change config, only view it)

eg.
```
username oxidize privilege 9 password 0 mysecretpassword
privilege exec level 9 show running-config view full
```
This user cannot 'config terminal', and the output of 'show run' is truncated.
```
#show running-config
Building configuration...

Current configuration : 53 bytes
!
boot-start-marker
boot-end-marker
!
!
!
!
!
!
end
```

Here's a not so helpful Cisco page sort of poorly explaining why 'show run' alone does not work.

http://www.cisco.com/c/en/us/support/docs/security-vpn/terminal-access-controller-access-control-system-tacacs-/23383-showrun.html

editing model/ios.rb to add 'view full' does the trick.

```
# diff -U1 /var/lib/gems/2.1.0/gems/oxidized-0.13.1/lib/oxidized/model/ios.rb /var/lib/gems/2.1.0/gems/oxidized-0.13.1/lib/oxidized/model/ios.rb~
--- /var/lib/gems/2.1.0/gems/oxidized-0.13.1/lib/oxidized/model/ios.rb	2016-04-08 08:09:49.289348269 -0700
+++ /var/lib/gems/2.1.0/gems/oxidized-0.13.1/lib/oxidized/model/ios.rb~	2016-04-06 17:40:16.885546258 -0700
@@ -42,3 +42,3 @@

-  cmd 'show running-config view full' do |cfg|
+  cmd 'show running-config' do |cfg|
     cfg = cfg.each_line.to_a[3..-1]
```


It seems to be just fine in my testing, so I would also advocate making it the default.

**Comment by ytti** [comment]
I fear lot of people are running platforms which won't accept 'show run view full' command. So I'm hesitant. We could do it conditionally based on some variable in config.

**Comment by jkrauska** [comment]
How difficult would it be to identify if the result of 'show running-config view full' was empty/invalid and then fall back or just run both all the time?

Racid added a patch for this in 2.3.7.
"rancid: add 'show running-config view full' for role cli - Pavel Korovin"

It appears to just run both commands in sequence? (including others which I know will fail on some of my systems)
from rancid 3.4.1
```
etc/rancid.types.base-cisco;command;ios::WriteTerm;more system:running-config;ASA/PIX
etc/rancid.types.base:cisco;command;ios::WriteTerm;show running-config view full;workaround for role-based CLI
etc/rancid.types.base-cisco;command;ios::WriteTerm;show running-config
```

I fear asking the user to add a special flag will be confusing for new users.

I realize 'lets do it because rancid does it' is not an awesome way to do a feature request.  Thanks for listening.

**Comment by ytti** [comment]
maybe we could do something ile

```
cmd 'show running-config view fill' do |cfg|
  @run_cfg = process_run(cfg)
end

cmd 'show running-config' do |cfg|
  if @run_cfg.lines > 5
    ""
 else
   process_run(cfg)
end


def process_run cfg
   ... old stuff from show running-config ...
end
```

So we first try to run view, process it normally. Then on normal viewless, we see if we've already gotten enough lines from view, if so, we return empty string, otherwise we process it normally.

I'm happy to accept PR for tested solution

**Comment by ytti** [comment]
Either proposed solution, or then vars ``ìos_rbac: true``` where we'd have in ios.rb:

```
show_run  = "show running-config"
show_run += " view full" if vars(:ios_rbac)
cmd show_run do |cfg|
  poop
end
```
